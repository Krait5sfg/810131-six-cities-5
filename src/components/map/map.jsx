import React, {PureComponent} from 'react';
import '../../../node_modules/leaflet/dist/leaflet.css';
import leaflet from 'leaflet';
import PropTypes from 'prop-types';
import {OfferPropTypes} from '../../utils/property-type';
import {TypePage} from '../../utils/const';

const ICON_PATH = `img/pin.svg`;
const AMSTERDAM_COORDINATES = [52.38333, 4.9];
const ID_MAP_CONTAINER = `map`;
const ZOOM = 12;
const IconSize = {
  WIDTH: 27,
  HEIGHT: 39
};

export default class Map extends PureComponent {

  componentDidMount() {
    this._setMap();
  }

  componentDidUpdate() {
    this._resetMap();
    this._setMap();
  }

  render() {
    const {typePage} = this.props;

    let elementClassName = ``;
    if (typePage === TypePage.MAIN) {
      elementClassName = `cities__map`;
    } else if (typePage === TypePage.OFFER) {
      elementClassName = `property__map`;
    }

    return <section id="map" className={`${elementClassName} map`}></section>;
  }

  _setMap() {
    const {offers} = this.props;
    const offerCoordinates = offers.map((offer) => offer.coordinates);

    // настройки leaflet
    const icon = leaflet.icon({
      iconUrl: ICON_PATH,
      iconSize: [IconSize.WIDTH, IconSize.HEIGHT]
    });

    this._map = leaflet.map(ID_MAP_CONTAINER, {
      center: AMSTERDAM_COORDINATES,
      zoom: ZOOM,
      zoomControl: false,
      marker: true
    });

    this._map.setView(AMSTERDAM_COORDINATES, ZOOM);

    leaflet
      .tileLayer(`https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
        attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
      })
      .addTo(this._map);

    // добавляет координаты на карту leaflet
    offerCoordinates.forEach((coordinates) => {
      leaflet
        .marker(coordinates, {icon})
        .addTo(this._map);
    });
  }

  _resetMap() {
    this._map.remove();
  }
}

Map.propTypes = {
  offers: PropTypes.arrayOf(OfferPropTypes).isRequired,
  typePage: PropTypes.string.isRequired
};
