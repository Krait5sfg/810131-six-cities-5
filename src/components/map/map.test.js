import React from 'react';
import renderer from 'react-test-renderer';
import {Map} from './map';

const offers = [
  {
    city: ``,
    accommodation: {
      bedroomsCount: 0,
      features: [],
      guestsLimit: 0,
      isPremium: true,
      price: 241,
      rating: 4.4,
      title: `The Joshua Tree House`,
      type: `room`,
    },
    cityLocation: {latitude: 0, longitude: 0, zoom: 0},
    coordinates: [0, 0],
    description: ``,
    host: {
      avatar: ``,
      isPro: true,
      name: ``,
    },
    id: 1,
    images: [],
    isFavorite: false,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/6.jpg`,
  },
  {
    city: ``,
    accommodation: {
      bedroomsCount: 0,
      features: [],
      guestsLimit: 0,
      isPremium: true,
      price: 241,
      rating: 4.4,
      title: `The Joshua Tree House`,
      type: `room`,
    },
    cityLocation: {latitude: 0, longitude: 0, zoom: 0},
    coordinates: [0, 0],
    description: ``,
    host: {
      avatar: ``,
      isPro: true,
      name: ``,
    },
    id: 2,
    images: [],
    isFavorite: false,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/6.jpg`,
  },
  {
    city: ``,
    accommodation: {
      bedroomsCount: 0,
      features: [],
      guestsLimit: 0,
      isPremium: true,
      price: 241,
      rating: 4.4,
      title: `The Joshua Tree House`,
      type: `room`,
    },
    cityLocation: {latitude: 0, longitude: 0, zoom: 0},
    coordinates: [0, 0],
    description: ``,
    host: {
      avatar: ``,
      isPro: true,
      name: ``,
    },
    id: 3,
    images: [],
    isFavorite: true,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/6.jpg`,
  },
];

describe(`Map render correctly`, () => {
  it(`Map on Main page`, () => {
    const tree = renderer
      .create(<Map
        offers={offers}
        typePage={`main`}
        idActiveCardForMap={10}
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Map on Main page without active card`, () => {
    const tree = renderer
      .create(<Map
        offers={offers}
        typePage={`main`}
        idActiveCardForMap={0}
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Map on Offer page`, () => {
    const tree = renderer
      .create(<Map
        offers={offers}
        typePage={`offer`}
        idActiveCardForMap={10}
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
