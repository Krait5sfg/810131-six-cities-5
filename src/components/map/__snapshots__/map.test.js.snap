// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map render correctly Map on Main page 1`] = `
<section
  className="cities__map map"
  id="map"
/>
`;

exports[`Map render correctly Map on Main page without active card 1`] = `
<section
  className="cities__map map"
  id="map"
/>
`;

exports[`Map render correctly Map on Offer page 1`] = `
<section
  className="property__map map"
  id="map"
/>
`;
